### PR TITLE
Fixing logging rotation

### PR DESCRIPTION
**Pull request by aleksandra-tarkowska**
This PR should resolve issue with logging rotation. It seems that standard logging is thread-safe, but the standard handlers cannot be safely used to log from multiple processes into a single file. Now it was replaced by ```ConcurrentLogHandler``` that allows shared access to the log file. More in https://pypi.python.org/pypi/ConcurrentLogHandler/0.9.1

To test:
 - ssh to trout.o.org
 - ```ls -al /opt/hudson/workspace/OMERO-5.1-breaking-deploy/src/dist/var/log/```
 - start browsing through web and watch how OMEROweb.log is growing. Once is 5MB it should rotate.
Hint: open large dataset and/or tiled image, they produce more logs ;-)

```
-rw-r--r--   1 ola  staff         0  1 Jul 12:43 OMEROweb.lock
-rw-r--r--   1 ola  staff     59957  1 Jul 12:48 OMEROweb.log
-rw-r--r--   1 ola  staff   5243139  1 Jul 12:48 OMEROweb.log.1
-rw-r--r--   1 ola  staff   5242906  1 Jul 12:46 OMEROweb.log.2
```

Note: number of files were extended to 10. Logs will rich max 50MB
cc: @manics @joshmoore @kennethgillen

### CODE REVIEW

File: components/tools/OmeroWeb/build.xml [hunk 1]
```diff
@@ -59,6 +59,21 @@
                 </touch>
             </then>
         </if>
+        <if><not><available file="${basedir}/../target/lib/python/clogging"/></not>
```
Comment (joshmoore) [inline components/tools/OmeroWeb/build.xml]: Might be worth discussing making this `lib/fallback` rather than `lib/python`

Comment (aleksandra-tarkowska) [inline components/tools/OmeroWeb/build.xml]: Is fallback in python path?

Comment (aleksandra-tarkowska) [inline components/tools/OmeroWeb/build.xml]: If that is a problem I can create local dir in omeroweb.

Comment (joshmoore) [inline components/tools/OmeroWeb/build.xml]: It's a problem, and both `lib/python` and `lib/fallback` are on `PYTHONPATH`. It's just that `fallback` is added to the end, so if someone wanted to update `ConcurrentLogHandler` locally, they could.

Comment (aleksandra-tarkowska) [inline components/tools/OmeroWeb/build.xml]: @joshmoore np, works fine as well

Comment (aleksandra-tarkowska) [inline components/tools/OmeroWeb/build.xml]: Do you prefer if all web dependencies, like django pipeline, etc. will go to fallback?

Comment (aleksandra-tarkowska) [inline components/tools/OmeroWeb/build.xml]: @joshmoore I am affraid web tests don't like fallback

```
[exec] E   ValueError: Unable to configure handler 'default': Cannot resolve 'clogging.cloghandler.ConcurrentRotatingFileHandler': No module named clogging
```

**Comment by aleksandra-tarkowska** [comment]
Travis hanging again?

**Comment by mtbc** [comment]
Observed a log rotation:
```
$ ls -l OMEROweb.log*
-rw-r--r-- 1 hudson hudson 1200280 Jul  2 13:31 OMEROweb.log
-rw-rw-r-- 1 hudson hudson 5245906 Jul  2 13:30 OMEROweb.log.1
```
and 5Mib is 5242880 bytes. :+1:

**Comment by mtbc** [comment]
No obvious problems with code changes and took quite some effort to get even that much of logs, so from my perspective this seems fine to merge.

**Comment by sbesson** [comment]
Reincluded into tomorrow's merge build.

**Comment by mtbc** [comment]
Limit to 11 log files seems to work: at rotation time,

* `OMEROweb.log.1` got replaced with `OMEROweb.log`
* `OMEROweb.log.10` was replaced with the earlier logfile
* no `OMEROweb.log.11` was created.

So this aspect works as expected too.

**Comment by aleksandra-tarkowska** [comment]
thanks @mtbc, it took a while to get that tested.

**Comment by aleksandra-tarkowska** [comment]
@joshmoore do we need to retest that on merge? Testing is quite time consuming.

on merge 
```
[atarkowska@ome-c6100-3 ~]$ ls -al /opt/hudson/workspace/OMERO-5.1-merge-deploy/src/dist/var/log/
total 2204
-rw-rw-r-- 1 hudson hudson       0 Jul  3 04:43 OMEROweb.lock
-rw-rw-r-- 1 hudson hudson   54011 Jul  3 09:24 OMEROweb.log
-rw-rw-r-- 1 hudson hudson       0 Jul  3 04:43 OMEROweb_request.lock
-rw-rw-r-- 1 hudson hudson       0 Jul  3 04:43 OMEROweb_request.log
```

**Comment by manics** [comment]
Could script it, e.g.:

`curl -b 'csrftoken=FtsysV2i1sxtgVdilpCbg6KcDG2RZ9iw; phpbb3_a9mfv_u=2883; phpbb3_a9mfv_k=9d511a6ca9788852; phpbb3_a9mfv_sid=de7375b0fa7f4105b8f443edd9e4a92b; style_cookie=null; _ga=GA1.2.2038438152.1432729355; sessionid_merge=p23xttqn048mpoq93p96mzbia0vcr33y' https://trout.openmicroscopy.org/merge/webclient/`

`-b` argument is the cookie obtained from browers web developer tools or similar after login.

**Comment by aleksandra-tarkowska** [comment]
@manics I had script, but the problem is small amount of logs produced by web when debug=false. It took half of the day to test it.

**Comment by manics** [comment]
It's rolled over at 5242947

```
[spli@ome-c6100-3 log]$ ls -l
total 426188
-rw-rw-r-- 1 hudson hudson 427209492 Jul  3 16:14 Blitz-0.log
-rw-rw-r-- 1 hudson hudson      1246 Jul  3 04:43 DropBox.log
-rw-rw-r-- 1 hudson hudson       114 Jul  3 04:43 FileServer.log
-rw-rw-r-- 1 hudson hudson    137471 Jul  3 15:43 Indexer-0.log
-rw-rw-r-- 1 hudson hudson         0 Jul  3 04:43 master.err
-rw-rw-r-- 1 hudson hudson         0 Jul  3 04:43 master.out
-rw-rw-r-- 1 hudson hudson      2053 Jul  3 04:43 MonitorServer.log
-rw-rw-r-- 1 hudson hudson         0 Jul  3 11:27 OMEROweb.lock
-rw-r--r-- 1 hudson hudson   3541559 Jul  3 16:13 OMEROweb.log
-rw-rw-r-- 1 hudson hudson   5242947 Jul  3 11:55 OMEROweb.log.1
-rw-rw-r-- 1 hudson hudson         0 Jul  3 11:27 OMEROweb_request.lock
-rw-rw-r-- 1 hudson hudson      2004 Jul  3 13:03 OMEROweb_request.log
-rw-rw-r-- 1 hudson hudson      6330 Jul  3 15:43 PixelData-0.log
-rw-rw-r-- 1 hudson hudson    227016 Jul  3 15:40 Processor-0.log
-rw-rw-r-- 1 hudson hudson       945 Jul  3 04:43 Tables-0.log
```

**Comment by joshmoore** [comment]
Do we know how this will fare on Windows?

**Comment by aleksandra-tarkowska** [comment]
Windows? What is that :-)

The University of Dundee is a registered Scottish Charity, No: SC015096

**Comment by joshmoore** [comment]
Not sure. I think it was a hot/late Friday afternoon delirium. Don't mind me.

**Comment by manics** [comment]
[OMERO-5.1-merge-deploy-win2012](https://ci.openmicroscopy.org/view/Failing/job/OMERO-5.1-merge-deploy-win2012/93/console) has been failing since Friday:
```
2015-07-06 04:48:42,805 [omego.upgrad] INFO  Stopping server
2015-07-06 04:48:42,805 [omego.extern] INFO  Running [old environment]: admin status --nodeonly
2015-07-06 04:48:42,805 [omego.extern] INFO  Executing [custom environment]: omero admin status --nodeonly
2015-07-06 04:48:44,071 [omego.extern] INFO  Running [old environment]: admin stop
2015-07-06 04:48:44,071 [omego.extern] INFO  Executing [custom environment]: omero admin stop
2015-07-06 04:48:57,573 [omego.upgrad] INFO  Stopping web
2015-07-06 04:48:57,573 [omego.upgrad] INFO  Removing web from IIS
2015-07-06 04:48:57,573 [omego.extern] INFO  Running [old environment]: web iis --remove
2015-07-06 04:48:57,573 [omego.extern] INFO  Executing [custom environment]: omero web iis --remove
Build timed out (after 30 minutes). Marking the build as aborted.
Build was aborted
```

**Comment by manics** [comment]
Hake merge is running, but `OMEROweb.log` and `OMEROweb_request.log` remain at 0 KB. But so does latest.

**Comment by aleksandra-tarkowska** [comment]
@joshmoore after an investigation @manics and I managed to start the latest build in ome-win2008. It required restart but we were not sure why exactly. Anyway logs file are created when you do ```bin\omero web iis (--remove)``` Unfortunately when iis is running we still get the following in EventViewer:
```
Failed to import callback module 'omero_web_iis'
The system cannot find the file specified.
```
and nothing is saved in logfiles. That seems to be unrelated to that PR as changing logger back doesn't help. The same issue is on windows latest.

**Comment by manics** [comment]
Rebooted ome-w2012r2, attempted to manually restart the merge OMERO.server. `omero web iis` is still hanging:
```
C:\Jenkins\workspace\OMERO-5.1-merge-deploy-win2012\OMERO.server-5.1.2-308-a7517
19-ice35-b550>bin\omero web iis
WARNING:root:Setting omero.web.force_script_name is deprecated, use omero.web.pr
efix
WARNING:root:Setting omero.web.force_script_name is deprecated, use omero.web.pr
efix
Post-processed 'omeroweb.viewer.min.css' as 'omeroweb.viewer.min.css'
Post-processed 'omeroweb.viewer.min.js' as 'omeroweb.viewer.min.js'

0 static files copied, 683 unmodified, 2 post-processed.
```

**Comment by manics** [comment]
Stopped the server, deleted `var/`, then in `omero shell`:
```
C:\Jenkins\workspace\OMERO-5.1-merge-deploy-win2012\OMERO.server-5.1.2-308-a7517
19-ice35-b550>bin\omero shell
WARNING: Readline services not available or not loaded.
WARNING: Proper color support under MS Windows requires the pyreadline library.
You can find it at:
http://ipython.org/pyreadline.html
Gary's readline needs the ctypes module, from:
http://starship.python.net/crew/theller/ctypes
(Note that ctypes is already part of Python versions 2.5 and newer).

Defaulting color scheme to 'NoColor'
Python 2.7.9 (default, Dec 10 2014, 12:28:03) [MSC v.1500 64 bit (AMD64)]
Type "copyright", "credits" or "license" for more information.

IPython 1.2.1 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: import omero_web_iis
WARNING:root:Setting omero.web.force_script_name is deprecated, use omero.web.pr
efix

In [2]: quit()
```
Importing the module creates `var/log/{OMEROweb.log, OMEROweb.lock, OMEROweb_request.log, OMEROweb_request.lock}`. The shell then hangs on `quit`.

**Comment by manics** [comment]
No idea if this is relevant (would require debugging django logging to see whether it's possible for `acquire()` to be called twice), but:
```
C:\Users\spli-x>\Jenkins\workspace\OMERO-5.1-merge-deploy-win2012\OMERO.server-5
.1.2-308-a751719-ice35-b550\bin\omero shell
WARNING: Readline services not available or not loaded.
WARNING: Proper color support under MS Windows requires the pyreadline library.
You can find it at:
http://ipython.org/pyreadline.html
Gary's readline needs the ctypes module, from:
http://starship.python.net/crew/theller/ctypes
(Note that ctypes is already part of Python versions 2.5 and newer).

Defaulting color scheme to 'NoColor'
Python 2.7.9 (default, Dec 10 2014, 12:28:03) [MSC v.1500 64 bit (AMD64)]
Type "copyright", "credits" or "license" for more information.

IPython 1.2.1 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: import cloghandler as cl

In [2]: f=cl.ConcurrentRotatingFileHandler('tmp.log')

In [3]: f
Out[3]: <cloghandler.ConcurrentRotatingFileHandler at 0x66be0f0>

In [4]: f.acquire()

In [5]: f.acquire()
```
Hangs at the second acquire. On OS X it doesn't.

**Comment by aleksandra-tarkowska** [comment]
@manics 
```
 In [1]: import cloghandler as cl
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
<ipython-input-1-8a1d0df364ab> in <module>()
----> 1 import cloghandler as cl

ImportError: No module named cloghandler

In [2]: from clogging.cloghandler import ConcurrentRotatingFileHandler

In [3]: f=ConcurrentRotatingFileHandler('tmp.log')

In [4]: f.acquire()

In [5]:
```
something wrong with PYTHONPATH?

**Comment by manics** [comment]
No sure what I did the first time. But calling `f.acquire()` twice still fails on Windows.

**Comment by aleksandra-tarkowska** [comment]
is that better if I add exclusion for windows and set as it was before? Unix seems to be fine

**Comment by manics** [comment]
:+1:

**Comment by aleksandra-tarkowska** [comment]
@manics @joshmoore that should exclude windows and use previous logging class I am will have a look at win2008 tomorrow

**Comment by manics** [comment]
https://ci.openmicroscopy.org/job/OMERO-5.1-merge-deploy-win2012/ is back now- I'm pretty sure `ConcurrentRotatingFileHandler` was causing python to hang on exit if `omero_web_iis` was imported. How about wait for a couple more builds/deploys, then merge on Monday?

**Comment by aleksandra-tarkowska** [comment]
@manics sure, not a problem.

**Comment by joshmoore** [comment]
All looks good, and as @aleksandra-tarkowska points out, this has been in for a few days without issue. One last question, though: I just saw:
```
OMEROweb.lock
```
are we going to have any issues on NFS?

**Comment by aleksandra-tarkowska** [comment]
This package bundles portalocker to deal with file locking, that only supports Unix (posix) an NT platforms.

**Comment by joshmoore** [comment]
OMERO bundles portalocker, too, so it's less an issue of which platforms are supported but which filesystems are supported. We've had substantial issues with locking on NFS, so if `var/log` is on NFS, what will happen?

**Comment by aleksandra-tarkowska** [comment]
@joshmoore could you describe how the set up should look exactly?

**Comment by joshmoore** [comment]
Whew. Ok. Finally managed to have something in place:

 * (root) added my dev directory to `/etc/exports`: "/opt/ome -mapall=jamoore"
 * (root) restarted NFS: `nfsd update`
 * (root) halted the lock server `killall -STOP rpc.lockd`
 * mounted the directory as myself: `mkdir /tmp/ome && mount_nfs localhost:/opt/ome /tmp/ome`
 * At this point, OMERO is pretty unusable if you try to start it from the mount: `/tmp/ome/dist/bin/omero admin restart` blocks with "nfs server localhost:/opt: lockd not responding". For me, this means that if the server is set up very poorly, it will be completely obvious to the user. (:+1:)
 * I then configured Python logging so that it would go to NFS. `bin/omero admin restart` and `bin/omero web start` worked fine. (minor issue: web shows no errors re: logging) The website is loadable (:+1:) though there is no logging, which is really a pretty minor problem since the logging before was also being lost.

**Comment by joshmoore** [comment]
@aleksandra-tarkowska : having looked into the `ConcurrentLogHandler` code, I was wondering if moving the single file (cloghandler.py) into omero_ext might be worth it? If so, could you cherry-pick https://github.com/joshmoore/openmicroscopy/commit/7d9d5cf03f6284e415f71950b7c5597982b43532 (optionally rebasing away the tar.gz)

**Comment by aleksandra-tarkowska** [comment]
@joshmoore my understanding from the past was that 3rd party libraries should be in lib/repository to simplify library update. If that has recently changed I can easily move that to omeroweb. Could you clarify that as you already asked me to move that to fallback https://github.com/openmicroscopy/openmicroscopy/pull/3922#discussion_r33670063?

**Comment by joshmoore** [comment]
Certainly for complicated libraries like django, the strategy you used makes complete sense. The choice of `lib/python` or `lib/fallback` is really just whether the user will want to update the library or not. But since `cloghandler` is so simple, putting it into `omero_ext` (as we did with portalocker, killableprocess, winprocess, argparse, etc) is *safer* since we're using our own namespace and there won't be any collision.

**Comment by joshmoore** [comment]
`OMEROweb.log` on trout today looks fine:

```
[jamoore@ome-c6100-3 log]$ ls -ltrad OMEROweb.*
-rw-rw-r-- 1 hudson hudson     0 Jul 27 04:41 OMEROweb.lock
-rw-rw-r-- 1 hudson hudson 22243 Jul 27 12:26 OMEROweb.log
```

**Pull request merged**